lib/msf/base/sessions/winrm_command_shell.rb

```diff
@@ -0,0 +1,149 @@
+
+    def shell_write(buf)
+      return unless buf
+
+      begin
+        framework.events.on_session_command(self, buf.strip)
+        shell.send_stdin(buf)
+      rescue ::Rex::SocketError, ::EOFError, ::IOError, ::Errno::EPIPE => e
+        shell_close
+        raise e
+      end
+    end
+
+    def cleanup
+      begin
+        stop_keep_alive_loop
+        shell.cleanup_shell
+      rescue WinRM::WinRMWSManFault => err
+        print_error('Shell may not have terminated cleanly')
+      end
+    end
+
+    ##
+    # :category: Msf::Session::Interactive implementors
+    #
+    def _interact_stream
+      fds = [user_input.fd]
+      while self.interacting
+        sd = Rex::ThreadSafe.select(fds, nil, fds, 0.5)
+        begin
+          @buffer_mutex.synchronize {
+            @buffer.each { |buf| user_output.print(buf) }
+            @buffer = []
+          }
+          if sd 
+            run_single((user_input.gets || '').chomp("\n"))
+            @check_stdin_event.signal()
+          end
+        rescue WinRM::WinRMWSManFault => err
+          print_error(err.fault_description)
+          shell_close
+        end
+        Thread.pass
+      end
+    end
+
+    def on_registered
+      start_keep_alive_loop
+    end
+
+    def start_keep_alive_loop
+      self.keep_alive_thread = framework.threads.spawn("WinRM-shell-keepalive", false, self.shell) do |thr_shell|
+        loop_delay = 0.5
+        loop do
+          tmp_buffer = []
+          output_seen = false
+          shell.read_stdout do |stdout, stderr|
+            if stdout or stderr
+              output_seen = true
+            end
+            tmp_buffer << stdout if stdout
+            tmp_buffer << stderr if stderr
+          end
+          @buffer_mutex.synchronize {
+            @buffer.concat(tmp_buffer)
+          }
+
+          # If our last request received stdout, let's be ready for some more
+          if output_seen
+            loop_delay = 0.5
+          else
+            # Gradual backoff
+            loop_delay *= 4
+            loop_delay = [loop_delay, 30].min
+          end
+
+          # Wait loop_delay seconds, or until an interactive thread wakes us up
+          @check_stdin_event.wait(loop_delay)
+          
+          Thread.pass
+        rescue WinRM::WinRMWSManFault => err
+          print_error(err.fault_description)
+          detected_shell_ended
+        rescue EOFError
+          # Shell has been terminated
+          detected_shell_ended
+        rescue Rex::HostUnreachable => err
+          detected_shell_ended(err.message)
+        rescue StandardError => err
+          detected_shell_ended(err.message)
+        end
+      end
+    end
+
+    def detected_shell_ended(reason="")
+      self.interacting = false
+      framework.events.on_session_interact_completed()
+      framework.sessions.deregister(self, reason)
+    end
+
+    def stop_keep_alive_loop
+      self.keep_alive_thread.kill
+    end
+
+    def tunnel_to_s
+      'WinRM'
```

Comment:
Could you please add the peer information in here? This typically has the local host/port information and the remote host/port information.

If `#tunnel_local` and `#tunnel_peer` are defined, then the `tunnel_to_s` will take over from within [`lib/msf/core/session.rb`](https://github.com/rapid7/metasploit-framework/blob/a7d255bbe5537822c614ede71933fdc6597dd369/lib/msf/core/session.rb#L135). The local port may change if a new socket is opened for each HTTP request but at least having the local IP, remote IP and port would be really helpful.

Reply:
Done - currently just using the same connection across requests.